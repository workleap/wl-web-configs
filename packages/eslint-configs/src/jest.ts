import type { Linter } from "eslint";
import jestPlugin from "eslint-plugin-jest";
import globals from "globals";
import type { ConfigWithExtends } from "./types.ts";

export interface JestConfigOptions {
    rules?: Partial<Linter.RulesRecord>;
}

export const jestGlobalIgnores = [];

export function jestConfig(options: JestConfigOptions = {}) {
    const {
        rules = {}
    } = options;

    return [{
        name: "@workleap/eslint-configs/jest",
        files: [
            "**/*.test.{js,jsx,ts,tsx}",
            "**/*-test.{js,jsx,ts,tsx}",
            "**/__tests__/*.{js,jsx,ts,tsx}",
            "**/test.{js,jsx,ts,tsx}"
        ],
        extends: [
            jestPlugin.configs["flat/recommended"]
        ],
        languageOptions: {
            globals: {
                ...globals.browser,
                ...globals.es2024,
                ...globals.node,
                ...globals.commonjs,
                ...globals.jest
            }
        },
        settings: {
            jest: {
                version: "detect"
            }
        },
        rules: {
            // Prefer spies to allow for automatic restoration.
            "jest/prefer-spy-on": "error",
            // Gives better failure messages for array checks.
            "jest/prefer-to-contain": "error",
            // Positioned last to allow the consumer to override any rules.
            ...rules
        }
    }] satisfies ConfigWithExtends[];
};
