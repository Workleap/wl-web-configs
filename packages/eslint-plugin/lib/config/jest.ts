import type { Linter } from "eslint";
import { concat } from "eslint-flat-config-utils";
import jestPlugin from "eslint-plugin-jest";
import { reactTestFiles, testFiles } from "../utils/patterns.ts";

const config: Linter.FlatConfig[] = await concat(
    {
        name: "workleap/jest",
        files: [...testFiles, ...reactTestFiles],
        ...jestPlugin.configs["flat/recommended"],
        rules: {
            // Prefer spies to allow for automatic restoration
            "jest/prefer-spy-on": "error",
            // Gives better failure messages for array checks
            "jest/prefer-to-contain": "error"
        }
    }
);

export default config;
