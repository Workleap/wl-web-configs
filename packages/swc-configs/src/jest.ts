import type { Config } from "@swc/core";

export interface DefineJestConfigOptions {
    react?: boolean;
    parser?: "ecmascript" | "typescript";
}

export function defineJestConfig(options: DefineJestConfigOptions = {}) {
    const {
        react = false,
        parser = "typescript"
    } = options;

    const config: Config = {
        jsc: {
            parser: parser === "ecmascript"
                ? {
                    syntax: "ecmascript",
                    jsx: react
                }
                : {
                    syntax: "typescript",
                    tsx: react
                },
            // The output environment that the code will be compiled for.
            target: "es2022",
            transform: react
                ? {
                    react: {
                        // Use "react/jsx-runtime".
                        runtime: "automatic",
                        // Use the native "Object.assign()" instead of "_extends".
                        useBuiltins: true
                    }
                }
                : undefined,
            // Import shims from an external module rather than inlining them in bundle files to greatly reduce the bundles size.
            // Requires to add "@swc/helpers" as a project dependency
            externalHelpers: true
        },
        module: {
            // The output module resolution system that the code will be compiled for.
            type: "es6",
            // Prevent SWC from exporting the `__esModule` property.
            strict: true,
            // Preserve dynamic imports.
            ignoreDynamic: true
        }
    };

    return config;
}
