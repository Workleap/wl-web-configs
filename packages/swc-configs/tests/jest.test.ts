import { defineJestConfig } from "../src/jest.ts";

describe("typescript parser", () => {
    test("when \"react\" is true, the react transform configuration is included", () => {
        const result = defineJestConfig({
            parser: "typescript",
            react: true
        });

        expect(result).toMatchSnapshot();
    });

    test("when \"react\" is false, the react transform configuration is not included", () => {
        const result = defineJestConfig({
            parser: "typescript",
            react: false
        });

        expect(result).toMatchSnapshot();
    });

    test("when \"react\" is not set, the react transform configuration is not included", () => {
        const result = defineJestConfig({
            parser: "typescript"
        });

        expect(result).toMatchSnapshot();
    });
});

describe("ecmascript parser", () => {
    test("when \"react\" is true, the react transform configuration is included", () => {
        const result = defineJestConfig({
            parser: "ecmascript",
            react: true
        });

        expect(result).toMatchSnapshot();
    });

    test("when \"react\" is false, the react transform configuration is not included", () => {
        const result = defineJestConfig({
            parser: "ecmascript",
            react: false
        });

        expect(result).toMatchSnapshot();
    });

    test("when \"react\" is not set, the react transform configuration is not included", () => {
        const result = defineJestConfig({
            parser: "ecmascript"
        });

        expect(result).toMatchSnapshot();
    });
});

test("when \"parser\" is not set, the configuration parser is typescript", () => {
    const result = defineJestConfig();

    expect(result).toMatchSnapshot();
});

