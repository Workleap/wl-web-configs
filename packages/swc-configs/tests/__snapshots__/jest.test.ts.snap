// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ecmascript parser when "react" is false, the react transform configuration is not included 1`] = `
{
  "jsc": {
    "externalHelpers": true,
    "parser": {
      "jsx": false,
      "syntax": "ecmascript",
    },
    "target": "es2022",
    "transform": undefined,
  },
  "module": {
    "ignoreDynamic": true,
    "strict": true,
    "type": "es6",
  },
}
`;

exports[`ecmascript parser when "react" is not set, the react transform configuration is not included 1`] = `
{
  "jsc": {
    "externalHelpers": true,
    "parser": {
      "jsx": false,
      "syntax": "ecmascript",
    },
    "target": "es2022",
    "transform": undefined,
  },
  "module": {
    "ignoreDynamic": true,
    "strict": true,
    "type": "es6",
  },
}
`;

exports[`ecmascript parser when "react" is true, the react transform configuration is included 1`] = `
{
  "jsc": {
    "externalHelpers": true,
    "parser": {
      "jsx": true,
      "syntax": "ecmascript",
    },
    "target": "es2022",
    "transform": {
      "react": {
        "runtime": "automatic",
        "useBuiltins": true,
      },
    },
  },
  "module": {
    "ignoreDynamic": true,
    "strict": true,
    "type": "es6",
  },
}
`;

exports[`typescript parser when "react" is false, the react transform configuration is not included 1`] = `
{
  "jsc": {
    "externalHelpers": true,
    "parser": {
      "syntax": "typescript",
      "tsx": false,
    },
    "target": "es2022",
    "transform": undefined,
  },
  "module": {
    "ignoreDynamic": true,
    "strict": true,
    "type": "es6",
  },
}
`;

exports[`typescript parser when "react" is not set, the react transform configuration is not included 1`] = `
{
  "jsc": {
    "externalHelpers": true,
    "parser": {
      "syntax": "typescript",
      "tsx": false,
    },
    "target": "es2022",
    "transform": undefined,
  },
  "module": {
    "ignoreDynamic": true,
    "strict": true,
    "type": "es6",
  },
}
`;

exports[`typescript parser when "react" is true, the react transform configuration is included 1`] = `
{
  "jsc": {
    "externalHelpers": true,
    "parser": {
      "syntax": "typescript",
      "tsx": true,
    },
    "target": "es2022",
    "transform": {
      "react": {
        "runtime": "automatic",
        "useBuiltins": true,
      },
    },
  },
  "module": {
    "ignoreDynamic": true,
    "strict": true,
    "type": "es6",
  },
}
`;

exports[`when "parser" is not set, the configuration parser is typescript 1`] = `
{
  "jsc": {
    "externalHelpers": true,
    "parser": {
      "syntax": "typescript",
      "tsx": false,
    },
    "target": "es2022",
    "transform": undefined,
  },
  "module": {
    "ignoreDynamic": true,
    "strict": true,
    "type": "es6",
  },
}
`;
